File: Server/www/framework/com/model/net/wrapper/update/UpdateNetWDirittiUtente.php
<?php

/*
 * PRApp  Copyright (C) 2019  Luca Bartolomei
 *
 * This file is part of PRApp.
 *
 *     PRApp is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     PRApp is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with PRApp.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace com\model\net\wrapper\update;

use InvalidArgumentException;
use com\model\db\enum\Diritto;
use com\model\net\wrapper\NetWrapper;
use com\model\db\wrapper\WDirittiUtente;
use com\model\net\serialize\ArrayDeserializable;
use com\model\net\wrapper\update\UpdateNetWDirittiUtente;

class UpdateNetWDirittiUtente implements NetWrapper
{

    /**
     * Metodo factory.
     *
     * @param int $idUtente
     * @param Diritto[] $diritti
     * @throws InvalidArgumentException
     * @return UpdateNetWDirittiUtente
     */
    private static function make($idUtente, $diritti)
    {
        if (is_null($idUtente) || is_null($diritti))
            throw new InvalidArgumentException("Uno o più parametri nulli");

        if (! is_int($idUtente) || ! is_array($diritti))
            throw new InvalidArgumentException("Uno o più parametri non del tipo giusto");

        if ($idUtente <= 0)
            throw new InvalidArgumentException("ID Utente non valido");

        foreach ($diritti as $diritto) {
            if (! ($diritto instanceof Diritto))
                throw new InvalidArgumentException("Uno o più parametri non del tipo giusto");
        }

        return new UpdateNetWDirittiUtente($idUtente, $diritti);
    }

    public static function of($array)
    {
        if (is_null($array) || ! is_array($array))
            throw new InvalidArgumentException("Array nullo o non valido.");

        if (! array_key_exists("idUtente", $array))
            throw new InvalidArgumentException("Dato id non trovato.");

        if (! array_key_exists("diritti", $array))
            throw new InvalidArgumentException("Dato id non trovato.");

        return self::make((int) $array["idUtente"], Diritto::ofArray($array["diritti"]));
    }

    /**
     * Identificativo dell'utente.
     *
     * @var int
     */
    private $idUtente;

    //Prima c'era idStaff: sostituito dallo staff selezionato

    /**
     * Diritti dell'utente nello staff.
     *
     * @var Diritto[]
     */
    private $diritti;

    private function __construct($idUtente, $idStaff, $diritti)
    {
        $this->idUtente = $idUtente;
        $this->idStaff = $idStaff;
        $this->diritti = $diritti;
    }

    /**
     *
     * @return int
     */
    public function getIdUtente()
    {
        return $this->idUtente;
    }

    /**
     *
     * @return Diritto[]
     */
    public function getDiritti()
    {
        return $this->diritti;
    }

    public function getWDirittiUtente($idStaff)
    {
        return WDirittiUtente::make(self::getIdUtente(), $idStaff, self::getDiritti());
    }
}

File: Android/PRApp/app/src/main/java/com/prapp/model/net/wrapper/update/UpdateNetWDirittiUtente.java
/*
 * PRApp  Copyright (C) 2019  Luca Bartolomei
 *
 * This file is part of PRApp.
 *
 *     PRApp is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     PRApp is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with PRApp.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.prapp.model.net.wrapper.update;

import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;
import com.prapp.model.db.enums.Diritto;
import com.prapp.model.net.Twinned;
import com.prapp.model.net.serialize.JSONSerializable;
import com.prapp.model.net.serialize.adapter.DirittoSetAdapter;
import com.prapp.model.net.wrapper.NetWrapper;

import java.util.Set;
import java.util.TreeSet;

public class UpdateNetWDirittiUtente implements Twinned, NetWrapper, JSONSerializable {

    public static final String CLASS_PATH = "com\\model\\net\\wrapper\\update\\UpdateNetWDirittiUtente";

    public UpdateNetWDirittiUtente getEmpty()
    {
        return new UpdateNetWDirittiUtente(0,0, new TreeSet<Diritto>());
    }

    @SerializedName("idUtente")
    private Integer idUtente;

    @SerializedName("idStaff")
    private Integer idStaff;

    @SerializedName("diritti")
    @JsonAdapter(DirittoSetAdapter.class)
    private Set<Diritto> diritti;

    public UpdateNetWDirittiUtente(Integer idUtente, Integer idStaff, Set<Diritto> diritti) {
        this.idUtente = idUtente;
        this.idStaff = idStaff;
        this.diritti = diritti;
    }

    public Integer getIdUtente() {
        return idUtente;
    }

    public Integer getIdStaff() {
        return idStaff;
    }

    public Set<Diritto> getDiritti() {
        return diritti;
    }

    @Override
    public String getRemoteClassPath() {
        return CLASS_PATH;
    }

}
